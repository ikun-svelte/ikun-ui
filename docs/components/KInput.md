---
title: KInput
lang: en-US
---

# KInput

Input data using keyboard.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/input
```

```bash [yarn]
yarn add @ikun-ui/input
```

```bash [npm]
npm install @ikun-ui/input
```

:::

## Basic usage

<demo src="input/basic.svelte" github="Input"></demo>

## Prefix icon and suffix icon input

Set the prefix through the `icon` attribute

<demo src="input/prefix.svelte" github="Input"></demo>

## Disabled input

Disable the input through the `disabled` attribute

<demo src="input/disabled.svelte" github="Input"></demo>

## Clearable

Make the Input clearable with the `clearable` attribute.

<demo src="input/clearable.svelte" github="Input"></demo>

## Password input

Enable the password input box and pass the `type` as `password`.

<demo src="input/password.svelte" github="Input"></demo>

## Prepend and append input

Use the `append` or `prepend` attribute to enable a button

<demo src="input/append.svelte" github="Input"></demo>

## Search input

Quickly implement a search box

::: tip
When search is `true` and you press `enter` or click the `prepend` button or click the `append` button, the event will be triggered.
In this case, the `triggerPrepend` event, `triggerPrepend` event and `enter` event will not be triggered.
:::

<demo src="input/search.svelte" github="Input"></demo>

## Sizes

Add `size` attribute to change the size of Input. It supports `sm`, `md` and `lg`.

<demo src="input/sizes.svelte" github="Input"></demo>

## Textarea

Setting the `autosize` prop for a textarea type of Input makes the height to automatically adjust based on the content.
An options object can be provided to `autosize` to specify the minimum and maximum number of lines the textarea can automatically adjust.

<demo src="input/textarea.svelte" github="Input"></demo>

## Input Props

| Name                | Type                                             | Default | Description                                                                    |
| ------------------- | ------------------------------------------------ | ------- | ------------------------------------------------------------------------------ |
| size                | `sm \| md \|lg`                                  | `md`    | size of Input.                                                                 |
| type                | `text \| password`                               | `text`  | Determine whether it is a password input                                       |
| value               | `string`                                         | `-`     | Binding value                                                                  |
| iconPrefix          | `string`                                         | `-`     | The class name of the prefix icon, following the unocss standard               |
| iconSuffix          | `string`                                         | `-`     | The class name of the suffix icon, following the unocss standard               |
| append              | `string`                                         | `-`     | The class name of the append button icon, following the unocss standard        |
| prepend             | `string`                                         | `-`     | The class name of the prepend button icon, following the unocss standard       |
| placeholder         | `string`                                         | `false` | Input's placeholder                                                            |
| disabled            | `boolean`                                        | `false` | Disable the Input                                                              |
| clearable           | `boolean`                                        | `false` | whether to show clear button, only works when type is not 'textarea'           |
| search              | `boolean`                                        | `false` | Enable `search` callback                                                       |
| useCompositionInput | `boolean`                                        | `false` | Bind value will be updated after the composition input ends                    |
| rows                | `number`                                         | `3`     | number of rows of `textarea`, only works when type is 'textarea'               |
| autosize            | `boolean \| { minRows: number: maxRows:number }` | `false` | whether `textarea` has an adaptive height, only works when type is 'textarea'. |
| cls                 | `string`                                         | `-`     | Additional class                                                               |
| attrs               | `Record<string, string>`                         | `{}`    | Additional attributes                                                          |

## Input Events

| Name             | Description                                                                                                                        | Type                                     |
| ---------------- | ---------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------- |
| input            | Event fired on input                                                                                                               | `(value: string)=> void`                 |
| enter            | Event fired when `enter` is pressed                                                                                                | `(e: Event)=> void`                      |
| keydown          | Event fired when `keyboard` is pressed                                                                                             | `(e: Event)=> void`                      |
| change           | Event fired when the `value` is changes                                                                                            | `(e: Event)=> void`                      |
| compositionstart | The compositionstart event is fired when a text composition system                                                                 | `(e: CompositionEvent)=> void`           |
| compositionend   | The compositionend event is fired when a text composition system                                                                   | `(e: CompositionEvent)=> void`           |
| compositionInput | Event fired when enable `useCompositionInput`                                                                                      | `(e: CompositionEvent)=> void`           |
| triggerPrepend   | Event fired on prepend button                                                                                                      | `(value: HTMLInputElement.value)=> void` |
| triggerAppend    | Event fired on append button                                                                                                       | `(value: HTMLInputElement.value)=> void` |
| search           | When search is true and you press `enter` or click the `prepend` button or click the `append` button, the event will be triggered. | `(value: HTMLInputElement.value)=> void` |

## Input Slots

| Name    | Description                     |
| ------- | ------------------------------- |
| prefix  | Customize input prefix content  |
| suffix  | Customize input suffix content  |
| append  | Customize input append content  |
| prepend | Customize input prepend content |
