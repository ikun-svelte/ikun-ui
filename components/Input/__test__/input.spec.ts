import { tick } from 'svelte';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { fireEvent } from '@testing-library/svelte';
import KInput from '../src';
import KInputSlots from './fixture/input.slots.svelte';
import KInputSlotsAppend from './fixture/input.slots.append.svelte';
import KInputBindValueUpdate from './fixture/input.bind.value.update.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KInput', () => {
	vi.mock('svelte', async () => {
		const actual = (await vi.importActual('svelte')) as object;
		return {
			...actual,
			// @ts-ignore
			onMount: (await import('svelte/internal')).onMount
		};
	});

	test('props: size', async () => {
		const sizes = ['sm', 'md', 'lg'];
		for (const size of sizes) {
			host.remove();
			initHost();
			const instance = new KInput({
				target: host,
				props: {
					size,
					iconPrefix: 'i-prefix',
					iconSuffix: 'i-suffix'
				}
			});
			expect(instance).toBeTruthy();
			expect((host as HTMLElement)!.innerHTML.includes(`k-input__${size}`)).toBeTruthy();
			expect(
				(host as HTMLElement)!.innerHTML.includes(`k-input--icon__${size} k-input--prefix-icon`)
			).toBeTruthy();
			expect(
				(host as HTMLElement)!.innerHTML.includes(`k-input--icon__${size} k-input--suffix-icon`)
			).toBeTruthy();
			expect(host.innerHTML).matchSnapshot();
		}
	});

	test('props: value', async () => {
		const instance = new KInput({
			target: host,
			props: {
				value: 'ikun'
			}
		});
		expect(instance).toBeTruthy();
		const inputElem = host.getElementsByTagName('input')[0];
		expect(inputElem.value).toBe('ikun');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: placeholder', async () => {
		const placeholder = '我见青山多妩媚，料青山见我应如是';
		const instance = new KInput({
			target: host,
			props: {
				placeholder
			}
		});
		expect(instance).toBeTruthy();
		const inputElm = host.getElementsByTagName('input')[0];
		expect(inputElm.placeholder).toEqual(placeholder);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		const instance = new KInput({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		const inputElm = host.getElementsByTagName('input')[0];
		expect(inputElm.disabled).not.toBeNull();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: iconPrefix', async () => {
		const instance = new KInput({
			target: host,
			props: {
				iconPrefix: 'i-carbon-logo-svelte'
			}
		});
		expect(instance).toBeTruthy();
		const icon = host.getElementsByClassName('k-input--icon')[0];
		expect(icon).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: iconSuffix', async () => {
		const instance = new KInput({
			target: host,
			props: {
				iconSuffix: 'i-carbon-search'
			}
		});
		expect(instance).toBeTruthy();
		const icon = host.getElementsByClassName('k-input--icon')[0];
		expect(icon).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: cls', async () => {
		const instance = new KInput({
			target: host,
			props: {
				cls: 'ikun'
			}
		});
		expect(instance).toBeTruthy();
		const inputElm = host.getElementsByClassName('k-input')[0];
		expect(inputElm.className.includes('ikun')).toBe(true);
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: password', async () => {
		const instance = new KInput({
			target: host,
			props: {
				type: 'password'
			}
		});
		expect(instance).toBeTruthy();
		let iconElm = host.querySelector('.i-carbon-view-off');
		expect(host.innerHTML.includes('type="password"')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-view-off')).toBeTruthy();
		iconElm.click();
		await tick();
		expect(host.innerHTML.includes('type="text"')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-view')).toBeTruthy();
		iconElm = host.querySelector('.i-carbon-view');
		iconElm.click();
		await tick();
		expect(host.innerHTML.includes('type="password"')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-view-off')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: prepend', async () => {
		const instance = new KInput({
			target: host,
			props: {
				prepend: 'i-carbon-logo-svelte'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('i-carbon-logo-svelte')).toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__right')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: append', async () => {
		const instance = new KInput({
			target: host,
			props: {
				append: 'i-carbon-logo-svelte'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('i-carbon-logo-svelte')).toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__left')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: append & prepend', async () => {
		const instance = new KInput({
			target: host,
			props: {
				append: 'i-carbon-logo-svelte',
				prepend: 'i-carbon-logo-vue'
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('i-carbon-logo-svelte')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-logo-vue')).toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__left')).not.toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__left')).not.toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: textarea & rows', async () => {
		const instance = new KInput({
			target: host,
			props: {
				type: 'textarea',
				rows: 6
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('style="min-height: -24px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: textarea & autosize', async () => {
		const instance = new KInput({
			target: host,
			props: {
				type: 'textarea',
				autosize: { minRows: 1, maxRows: 2 }
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.innerHTML.includes('style="min-height: -4px; height: -8px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: attrs', async () => {
		const objAttr = {
			name: 'KInput'
		};
		const arrAttr = ['i', 'kun'];
		const instance = new KInput({
			target: host,
			props: {
				attrs: {
					type: 'number',
					strAttr: '你干嘛,哎哟',
					numAttr: 8,
					nullAttr: null,
					undefinedAttr: undefined,
					objAttr,
					arrAttr
				}
			}
		});
		expect(instance).toBeTruthy();
		const inputElm = host.getElementsByTagName('input')[0];
		expect(inputElm.type).toBe('number');
		expect(inputElm.getAttribute('strAttr')).toBe('你干嘛,哎哟');
		expect(inputElm.getAttribute('numAttr')).toBe('8');
		expect(inputElm.getAttribute('nullAttr')).toBeNull();
		expect(inputElm.getAttribute('undefinedAttr')).toBeNull();
		expect(inputElm.getAttribute('objAttr')).not.toEqual(objAttr);
		expect(inputElm.getAttribute('objAttr')).toEqual(objAttr.toString());
		expect(inputElm.getAttribute('arrAttr')).not.toEqual(arrAttr);
		expect(inputElm.getAttribute('arrAttr')).toEqual(arrAttr.toString());
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger input event', async () => {
		let value = '';
		const mockFn = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				value
			}
		});
		await tick();
		instance.$on('input', (v) => {
			value = v.detail;
			mockFn();
		});
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.value = 'ikun';
		inputElm.dispatchEvent(new Event('input'));
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).toBeCalled();
		expect(value).toBe('ikun');
	});

	test('event: should trigger composition input event', async () => {
		let value = '';
		const mockInputFn = vi.fn();
		const mockCompositionInputFn = vi.fn();
		const mockCompositionStart = vi.fn();
		const mockCompositionEnd = vi.fn();

		const instance = new KInput({
			target: host,
			props: {
				value,
				useCompositionInput: true
			}
		});
		await tick();
		instance.$on('input', () => {
			mockInputFn();
		});

		instance.$on('compositionstart', () => {
			mockCompositionStart();
		});

		instance.$on('compositionend', () => {
			mockCompositionEnd();
		});

		instance.$on('compositionInput', (v) => {
			value = v.detail;
			mockCompositionInputFn();
		});

		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.value = 'ikun';
		inputElm.dispatchEvent(new Event('compositionstart'));
		inputElm.dispatchEvent(new Event('input'));
		await tick();
		expect(instance).toBeTruthy();
		expect(mockCompositionStart).toBeCalled();
		expect(mockInputFn.mock.calls.length).toBe(1);
		expect(mockCompositionInputFn.mock.calls.length).toBe(0);
		expect(value).toBe('');

		inputElm.value = 'ikun ikun';
		inputElm.dispatchEvent(new Event('compositionend'));
		await tick();
		expect(mockCompositionEnd).toBeCalled();
		expect(mockInputFn.mock.calls.length).toBe(2);
		expect(mockCompositionInputFn.mock.calls.length).toBe(1);
		expect(value).toBe('ikun ikun');
	});

	test('event: should trigger enter event when pressing enter', async () => {
		const mockFn = vi.fn();
		const instance = new KInput({
			target: host
		});
		await tick();
		instance.$on('enter', mockFn);
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.dispatchEvent(
			new KeyboardEvent('keydown', {
				bubbles: true,
				cancelable: true,
				key: 'Enter',
				code: 'Enter'
			})
		);
		await tick();
		expect(mockFn).toBeCalled();
	});

	test('event: should trigger search event when pressing enter', async () => {
		const mockFn = vi.fn();
		const mockSearchFn = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				search: true
			}
		});
		await tick();
		instance.$on('enter', mockFn);
		instance.$on('search', mockSearchFn);
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.dispatchEvent(
			new KeyboardEvent('keydown', {
				bubbles: true,
				cancelable: true,
				key: 'Enter',
				code: 'Enter'
			})
		);
		await tick();
		expect(mockFn).not.toBeCalled();
		expect(mockSearchFn).toBeCalled();
	});

	test('event: should trigger keydown event when not pressing enter', async () => {
		const mockFn = vi.fn();
		const instance = new KInput({
			target: host
		});
		await tick();
		instance.$on('enter', mockFn);
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.dispatchEvent(
			new KeyboardEvent('keydown', {
				bubbles: true,
				cancelable: true,
				key: 'A',
				code: 'KeyA'
			})
		);
		await tick();
		expect(mockFn).not.toBeCalled();
	});

	test('event: should trigger change event', async () => {
		let value = '';
		const mockFn = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				value
			}
		});
		await tick();
		instance.$on('change', (v) => {
			value = (v.detail.target as HTMLInputElement).value;
			mockFn();
		});
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.value = 'ikun';
		inputElm.dispatchEvent(new Event('change'));
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).toBeCalled();
		expect(inputElm).toBeTruthy();
		expect(value).toEqual('ikun');
	});

	test('event: should not trigger event when disabled', async () => {
		const mockFn = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		await tick();
		instance.$on('input', mockFn);
		instance.$on('change', mockFn);
		instance.$on('enter', mockFn);
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.dispatchEvent(new Event('input'));
		inputElm.dispatchEvent(
			new KeyboardEvent('keydown', {
				bubbles: true,
				cancelable: true,
				key: 'Enter',
				code: 'KeyEnter'
			})
		);
		inputElm.value = 'ikun';

		inputElm.dispatchEvent(new Event('change'));
		await tick();
		expect(mockFn).not.toBeCalled();
	});

	test('event: should trigger prepend & append click', async () => {
		let value = '';
		let valueClick = '';
		const mockFn = vi.fn();
		const mockFnClick = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				value,
				append: 'i-carbon-logo-svelte',
				prepend: 'i-carbon-logo-vue'
			}
		});
		await tick();
		instance.$on('input', (v) => {
			value = v.detail;
			mockFn();
		});
		instance.$on('triggerPrepend', (v) => {
			valueClick = v.detail;
			mockFnClick();
		});
		instance.$on('triggerAppend', (v) => {
			valueClick = v.detail;
			mockFnClick();
		});
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.value = 'ikun';
		inputElm.dispatchEvent(new Event('input'));
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).toBeCalled();
		expect(value).toBe('ikun');
		const btnAppend = host.querySelector('.k-input--append');
		const btnPrepend = host.querySelector('.k-input--prepend');
		btnAppend.click();
		await tick();
		expect(mockFnClick).toBeCalled();
		expect(valueClick).toBe('ikun');
		valueClick = '';
		btnPrepend.click();
		await tick();
		expect(mockFnClick).toBeCalledTimes(2);
		expect(valueClick).toBe('ikun');
	});

	test('event: should trigger prepend & append search', async () => {
		let value = '';
		let valueClick = '';
		const mockFn = vi.fn();
		const mockFnClick = vi.fn();
		const mockFnSearch = vi.fn();
		const instance = new KInput({
			target: host,
			props: {
				value,
				append: 'i-carbon-logo-svelte',
				prepend: 'i-carbon-logo-vue',
				search: true
			}
		});
		await tick();
		instance.$on('input', (v) => {
			value = v.detail;
			mockFn();
		});
		instance.$on('triggerPrepend', () => {
			mockFnClick();
		});
		instance.$on('triggerAppend', () => {
			mockFnClick();
		});
		instance.$on('search', (v) => {
			valueClick = v.detail;
			mockFnSearch();
		});
		const inputElm = host.getElementsByTagName('input')[0];
		inputElm.value = 'ikun';
		inputElm.dispatchEvent(new Event('input'));
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).toBeCalled();
		expect(value).toBe('ikun');
		const btnAppend = host.querySelector('.k-input--append');
		const btnPrepend = host.querySelector('.k-input--prepend');
		btnAppend.click();
		await tick();
		expect(mockFnClick).not.toBeCalled();
		expect(mockFnSearch).toBeCalled();
		expect(valueClick).toBe('ikun');
		valueClick = '';
		btnPrepend.click();
		await tick();
		expect(mockFnClick).not.toBeCalled();
		expect(mockFnSearch).toBeCalledTimes(2);
		expect(valueClick).toBe('ikun');
	});

	test('slot: prefix and suffix', async () => {
		const instance = new KInputSlots({
			target: host
		});
		expect(instance).toBeTruthy();
		const prefixElm = host.getElementsByClassName('prefix')[0];
		const suffixElm = host.getElementsByClassName('suffix')[0];
		expect(prefixElm).toBeTruthy();
		expect(suffixElm).toBeTruthy();
	});

	test('slot: prepend and append', async () => {
		const instance = new KInputSlotsAppend({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-logo-svelte')).toBeTruthy();
		expect(host.innerHTML.includes('i-carbon-logo-vue')).toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__left')).not.toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded__left')).not.toBeTruthy();
		expect(host.innerHTML.includes('k-input__rounded')).not.toBeTruthy();
	});

	test('props: bind value be updated in real time', async () => {
		const instance = new KInputBindValueUpdate({
			target: host
		});
		expect(instance).toBeTruthy();

		const spanElm = host.querySelector('span')!;
		expect(spanElm).toBeTruthy();
		expect(spanElm.textContent).toBe('');

		const inputElm = host.querySelector('input')!;
		expect(inputElm).toBeTruthy();

		await fireEvent.input(inputElm, { target: { value: 'input change' } });
		await tick();

		expect(inputElm.value).toBe('input change');

		expect(spanElm.textContent).toBe('input change');
	});

	test('props: clearable', async () => {
		const instance = new KInput({
			target: host,
			props: {
				clearable: true,
				value: ''
			}
		});
		expect(instance).toBeTruthy();
		const inputElm = host.querySelector('input');
		expect(host.querySelector('.k-input--clear-icon')).toBeNull();
		instance.$set({ value: 'ikun-ui' });
		await fireEvent.focus(inputElm);
		await tick();
		expect(host.querySelector('.k-input--clear-icon')).toBeTruthy();
		const clearElm = host.querySelector('.k-input--clear-icon');
		clearElm.click();
		await tick();
		expect(inputElm?.value).toBe('');
		expect(host.innerHTML).matchSnapshot();
	});
});
