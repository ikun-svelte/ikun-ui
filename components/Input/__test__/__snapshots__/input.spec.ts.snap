// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KInput > props: append & prepend 1`] = `"<div class=\\"k-input\\"><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--main__fill k-button--main__hover__fill k-button--main__active k-button--main__focus k-input--prepend k-input--prepend__md\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-logo-vue k-button--main__icon k-button--main__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>  </button> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> <button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--main__fill k-button--main__hover__fill k-button--main__active k-button--main__focus k-input--append k-input--append__md\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-logo-svelte k-button--main__icon k-button--main__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>  </button></div> "`;

exports[`Test: KInput > props: append 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded__left\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> <button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--main__fill k-button--main__hover__fill k-button--main__active k-button--main__focus k-input--append k-input--append__md\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-logo-svelte k-button--main__icon k-button--main__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>  </button></div> "`;

exports[`Test: KInput > props: attrs 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"number\\" placeholder=\\"\\" strattr=\\"你干嘛,哎哟\\" numattr=\\"8\\" objattr=\\"[object Object]\\" arrattr=\\"i,kun\\">    </div> </div> "`;

exports[`Test: KInput > props: clearable 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;

exports[`Test: KInput > props: cls 1`] = `"<div class=\\"k-input ikun\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;

exports[`Test: KInput > props: disabled 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__disabled k-input__disabled__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input__disabled k-input__disabled__dark\\" disabled=\\"\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;

exports[`Test: KInput > props: iconPrefix 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-logo-svelte k-input--icon k-input--icon__md k-input--prefix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;

exports[`Test: KInput > props: iconSuffix 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-carbon-search k-input--icon k-input--icon__md k-input--suffix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span></div> </div> "`;

exports[`Test: KInput > props: password 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"password\\" placeholder=\\"\\">  <div role=\\"button\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon--base__dark k-icon--role-button k-icon-transition i-carbon-view-off k-input--icon k-input--icon__md ml-1\\" role=\\"button\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span></div>  </div> </div> "`;

exports[`Test: KInput > props: placeholder 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"我见青山多妩媚，料青山见我应如是\\">    </div> </div> "`;

exports[`Test: KInput > props: prepend 1`] = `"<div class=\\"k-input\\"><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--main__fill k-button--main__hover__fill k-button--main__active k-button--main__focus k-input--prepend k-input--prepend__md\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-logo-svelte k-button--main__icon k-button--main__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>  </button> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded__right\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;

exports[`Test: KInput > props: size 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__sm k-input__dark k-input__hover k-input__focus k-input__rounded\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-prefix k-input--icon k-input--icon__sm k-input--prefix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-suffix k-input--icon k-input--icon__sm k-input--suffix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span></div> </div> "`;

exports[`Test: KInput > props: size 2`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-prefix k-input--icon k-input--icon__md k-input--prefix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-suffix k-input--icon k-input--icon__md k-input--suffix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span></div> </div> "`;

exports[`Test: KInput > props: size 3`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__lg k-input__dark k-input__hover k-input__focus k-input__rounded\\"><span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-prefix k-input--icon k-input--icon__lg k-input--prefix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    <span class=\\"k-icon--base k-icon--base__dark k-icon-transition i-suffix k-input--icon k-input--icon__lg k-input--suffix-icon\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 24px; height: 24px;\\"></span></div> </div> "`;

exports[`Test: KInput > props: textarea & autosize 1`] = `" <div class=\\"k-input\\"><div class=\\"k-input--base k-input__dark k-input__hover k-input__focus k-input__rounded\\"><textarea class=\\"k-input--inner k-input--inner__textarea k-input--inner__dark\\" placeholder=\\"\\" style=\\"min-height: -4px; height: -8px;\\"></textarea></div></div>"`;

exports[`Test: KInput > props: textarea & rows 1`] = `" <div class=\\"k-input\\"><div class=\\"k-input--base k-input__dark k-input__hover k-input__focus k-input__rounded\\"><textarea class=\\"k-input--inner k-input--inner__textarea k-input--inner__dark\\" placeholder=\\"\\" style=\\"min-height: -24px;\\"></textarea></div></div>"`;

exports[`Test: KInput > props: value 1`] = `"<div class=\\"k-input\\"> <div class=\\"k-input--base k-input__md k-input__dark k-input__hover k-input__focus k-input__rounded\\"> <input class=\\"k-input--inner k-input--inner__dark\\" type=\\"text\\" placeholder=\\"\\">    </div> </div> "`;
